pkg/telemetryquerylanguage/functions/tqlotel/func_replace_all_patterns.go

```diff
@@ -23,23 +23,44 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/telemetryquerylanguage/tql"
 )
 
-func ReplaceAllPatterns(target tql.GetSetter, regexPattern string, replacement string) (tql.ExprFunc, error) {
+const (
+	modeKey   = "key"
+	modeValue = "value"
+)
+
+func ReplaceAllPatterns(target tql.GetSetter, mode string, regexPattern string, replacement string) (tql.ExprFunc, error) {
 	compiledPattern, err := regexp.Compile(regexPattern)
 	if err != nil {
 		return nil, fmt.Errorf("the regex pattern supplied to replace_all_patterns is not a valid pattern: %w", err)
 	}
+	if mode != modeValue && mode != modeKey {
+		return nil, fmt.Errorf("invalid mode %v, must be either 'key' or 'value'", mode)
+	}
+
 	return func(ctx tql.TransformContext) interface{} {
 		val := target.Get(ctx)
 		if val == nil {
 			return nil
 		}
 		if attrs, ok := val.(pcommon.Map); ok {
 			updated := pcommon.NewMap()
-			attrs.CopyTo(updated)
-			updated.Range(func(key string, value pcommon.Value) bool {
-				stringVal := value.StringVal()
-				if compiledPattern.MatchString(stringVal) {
-					value.SetStringVal(compiledPattern.ReplaceAllLiteralString(stringVal, replacement))
+			updated.EnsureCapacity(attrs.Len())
```

Comment:
I think it will be a fairly common occurrence that this function does no useful work -- that none of the attributes match the given pattern. Consequently, it might make sense, then, to do one of two things:
* set a flag when the match succeeds, and only do the `target.Set` if it has succeeded at least once; this avoids rewriting target, although it doesn't avoid the allocation
* do a pre-pass over the keys looking for a match, and if none is found exit without allocating the `updated` value

Is it worth it?

Reply:
We haven't done serious benchmarks for OTTl or the transform processor yet, but we should. As both mature I say we should take performance improvements wherever we can find them.

Reply:
Although I think the suggestion can be done in a future PR if you open an issue to call it out. 

I think I'd also like to merge all the Replace functions into one.